From 1a881ee3853001a3e75c8cb69cc183611d4a4373 Mon Sep 17 00:00:00 2001
From: Alexandros Frantzis <alexandros.frantzis@collabora.com>
Date: Wed, 23 Jun 2021 17:22:20 +0300
Subject: [PATCH 002/147] user32: Expose function to send internal thread
 messages.

Expose a function to drivers so that they can send internal messages to
specific threads, without the need to target a specific HWND.

Signed-off-by: Alexandros Frantzis <alexandros.frantzis@collabora.com>
---
 dlls/user32/message.c   | 13 +++++++++++++
 dlls/user32/user32.spec |  1 +
 include/winuser.h       |  3 +++
 3 files changed, 17 insertions(+)

diff --git a/dlls/user32/message.c b/dlls/user32/message.c
index 44b08da79dd..dff2429e6e5 100644
--- a/dlls/user32/message.c
+++ b/dlls/user32/message.c
@@ -3388,6 +3388,19 @@ LRESULT MSG_SendInternalMessageTimeout( DWORD dest_pid, DWORD dest_tid,
     return ret;
 }
 
+/***********************************************************************
+ *		__wine_send_internal_message_timeout  (USER32.@)
+ *
+ * Same as SendMessageTimeoutW but sends the message to a specific thread
+ * without requiring a window handle. Only works for internal Wine messages.
+ */
+LRESULT CDECL __wine_send_internal_message_timeout( DWORD dest_pid, DWORD dest_tid,
+                                                    UINT msg, WPARAM wparam, LPARAM lparam,
+                                                    UINT flags, UINT timeout, PDWORD_PTR res_ptr )
+{
+    return MSG_SendInternalMessageTimeout( dest_pid, dest_tid, msg, wparam, lparam,
+                                           flags, timeout, res_ptr );
+}
 
 /***********************************************************************
  *		SendMessageTimeoutW  (USER32.@)
diff --git a/dlls/user32/user32.spec b/dlls/user32/user32.spec
index c963dbaaee1..d46768e6e58 100644
--- a/dlls/user32/user32.spec
+++ b/dlls/user32/user32.spec
@@ -839,3 +839,4 @@
 @ cdecl __wine_send_input(long ptr ptr)
 @ cdecl __wine_set_pixel_format(long long)
 @ cdecl __wine_set_user_driver(ptr long)
+@ cdecl __wine_send_internal_message_timeout(long long long long long long long ptr)
diff --git a/include/winuser.h b/include/winuser.h
index 1d37bd44344..fb7304a08dd 100644
--- a/include/winuser.h
+++ b/include/winuser.h
@@ -4408,6 +4408,9 @@ WORD        WINAPI SYSTEM_KillSystemTimer( WORD );
 
 #ifdef __WINESRC__
 WINUSERAPI BOOL CDECL __wine_send_input( HWND hwnd, const INPUT *input, const RAWINPUT *rawinput );
+LRESULT CDECL __wine_send_internal_message_timeout( DWORD dest_pid, DWORD dest_tid,
+                                                    UINT msg, WPARAM wparam, LPARAM lparam,
+                                                    UINT flags, UINT timeout, PDWORD_PTR res_ptr );
 
 /* Uxtheme hook functions and struct */
 
-- 
2.34.1

